src/plugins/keras_support/categorical_crossentropy_operation.cpp

```diff
@@ -0,0 +1,291 @@
+// Copyright (c) 2018 Hartmut Kaiser
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#include <phylanx/config.hpp>
+#include <phylanx/ir/node_data.hpp>
+#include <phylanx/execution_tree/primitives/node_data_helpers.hpp>
+#include <phylanx/plugins/keras_support/categorical_crossentropy_operation.hpp>
+#include <phylanx/util/blaze_traits.hpp>
+#include <phylanx/plugins/statistics/statistics_base.hpp>
+#include <phylanx/plugins/statistics/sum_operation.hpp>
+
+#include <hpx/include/lcos.hpp>
+#include <hpx/include/naming.hpp>
+#include <hpx/include/util.hpp>
+#include <hpx/throw_exception.hpp>
+#include <hpx/util/assert.hpp>
+
+#include <algorithm>
+#include <cstddef>
+#include <cstdint>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+#include <type_traits>
+
+#include <blaze/Math.h>
+#if defined(PHYLANX_HAVE_BLAZE_TENSOR)
+#include <blaze_tensor/Math.h>
+#endif
+
+///////////////////////////////////////////////////////////////////////////////
+namespace phylanx { namespace execution_tree { namespace primitives
+{
+    ///////////////////////////////////////////////////////////////////////////
+    match_pattern_type const cat_cross_operation::match_data =
+    {
+        hpx::util::make_tuple("categorical_crossentropy",
+        std::vector<std::string>{
+            "categorical_crossentropy(_1,_2)",
+            "categorical_crossentropy(_1,_2,_3)"
+        },
+        &create_cat_cross_operation, &create_primitive<cat_cross_operation>,
+        R"(target, output, from_logits
+        Args:
+
+            target (array_like) : input array
+            output (array_like) : output array
+            from_logits (boolean): boolean value
+
+        Returns:
+            The value should be the same as would be returned by the following
+            Python function:
+
+            def cat_cross(target, output, from_logits=False):
+                if from_logits:
+                    output = softmax(output)
+                else:
+                    output /= output.sum(axis=-1, keepdims=True)
+                output = np.clip(output, 1e-7, 1 - 1e-7)
+                return np.sum(target * -np.log(output), axis=-1, keepdims=False)
+        )")
+    };
+    const double clip_low = 1e-7;
+    const double clip_high = 1 - clip_low;
+
+    ///////////////////////////////////////////////////////////////////////////
+    cat_cross_operation::cat_cross_operation(primitive_arguments_type&& operands,
+        std::string const& name, std::string const& codename)
+      : primitive_component_base(std::move(operands), name, codename)
+    {}
+
+    primitive_argument_type cat_cross_operation::cat_cross0d(
+        arg_type&& target,arg_type&& output,bool from_logits) const
+    {
+        double v = 1;
+
+        if(!from_logits)
+            v = output.scalar()/output.scalar();
```

Comment:
Isn't the result of this computation always `1.0`?

Reply:
Not when output.scalar() is zero.

Reply:
Ok, in this case, wouldn't it be more explicit to write this as:
```
double v = 1;
if (!from_logits && output.scalar() == 0)
    v = std::numeric_limits<double>::quiet_NaN();
```
?